nachomemes/bot.py

```diff
@@ -46,19 +46,104 @@
 # recent meme requests (and the resulting meme message)
 RECENT: SimpleCache[int,Message] = SimpleCache(200)
 
-async def report(ctx: Union[Context,Message], ex: Exception, message: str="An error has occured"):
+def print_all_templates(config: GuildConfig) -> dict:
+    memes = STORE.list_memes(config.guild_id, ("name", "description"))
+    return {"embed": Embed(
+        title="Available Templates",
+        description = "".join((f"\n{m['name']}: *{m['description']}*" for m in memes))
+    )}
+
+def print_matching_templates(config: GuildConfig, template_name: str) -> dict:
+    memes = STORE.close_matches(config.guild_id, template_name, ("name", "description"))
+    return {"embed": Embed(
+        title="Matching templates",
+        description = "".join(f"\n{m['name']}: *{m['description']}*" for m in memes)
+    )}
+
+def print_template(config: GuildConfig, template_name: str) -> dict:
+    template = STORE.best_match(config.guild_id, template_name)
+    return {"embed": Embed(
+        title="Template info",
+        description = dedent(f"""\
+        Name: {template.name}
+        Description: *{template.description}*
+        Times used: {template.usage}
+        Expects {len(template.textboxes)} strings
+        Read more: {template.docs}""")
+    )}
+
+
+async def generate(config: GuildConfig, member: Member, data: Iterable[str]) -> dict:
+    buffer: Optional[BufferedIOBase] = None
+    try:
+        if not config.can_use(member):
+            return {"embed": Embed(
+                title="NO",
+                description = no_memes(config, member)
+            )}
+        if not data:
+            return print_all_templates(config)
+
+        (template_name, *text) = data
+        if not text:
+            if "*" in template_name:
+                return print_matching_templates(config, template_name)
+            else:
+                return print_template(config, template_name)
+
+        template = STORE.best_match(config.guild_id, template_name, True)
+        name = re.sub(r"\W+", "", str(text))
```

Comment:
Can we substitute dashes in here? I think it would be more readable.

Reply:
I guess.  I mean it's for a filename, it doesn't need to be readable or anything like that.  It just has to exist.

Reply:
It pops up in notifications where you can't see the actual image so I think it's worth.

Reply:
k